Add Granular Pitch Shifter effect

LostRobotMusic opened the pull request:
![image](https://github.com/LMMS/lmms/assets/34612565/6b94c4ab-e955-4fe5-8c88-efd291904a87)

Finally, after twenty year-sized years, LMMS has a pitch shifter that isn't garbage.  Yay!

The GUI was created by thismoon.

Almost all granular pitch shifters out there make the same fundamental mistake, that being the use of equal-gain fades instead of equal-power fades.  It is for this reason that this type of pitch shifter is infamous for only being useful for sound design purposes rather than high-quality CPU-efficient realtime pitch shifting.  It's a tad sad to watch many developers of these pitch shifters even go as far as to use autocorrelation algorithms and the like to try to correlate the grains, in a desperate attempt to help their equal-gain fades sound a bit less awful, with... varying levels of success... usually not much.

This pitch shifter doesn't make those fundamental mistakes.  By default, the fades used in this plugin are equal-power.  I explain why these are the correct fades in this video here: https://www.youtube.com/watch?v=-5cB3rec2T0

Along with this comes several features to help with ensuring the grains are uncorrelated so this grain shape is always optimal.  The simplest and subtlest way to achieve this is to use a very tiny amount of "Spray" which slightly randomizes the positions at which the grains are started from in the delay buffer.  This is done by default.

-----------------------

This pitch shifter comes with a "prefilter" lowpass which almost entirely eliminates any aliasing generated by the pitch shifting, _without_ needing to resort to oversampling.  This is possible by applying the filter _before_ the pitch shifting, slightly below the frequency that will be shifted up to the nyquist frequency.

While most pitch shifters have a fixed amount of latency, this one (optionally) dynamically changes its latency to the absolute lowest it can possibly be for your current settings.  This can be overridden by increasing the latency knob, which will set a minimum latency requirement.  This also controls the delay length of the feedback, which allows shifting the audio repeatedly as it is fed back through the delay buffer for infinity.

For those who would like to use this pitch shifter for more creative sound design purposes, the Random section covers this very nicely.  Spray can randomize the read locations of each grain up to quite a long while back in time.  Jitter randomizes the pitch of each grain, while Twitch randomizes when these grains are created.  The three of these together allow particularly satisfying buffer-shifting sounds, which will always sound perfectly clean due to the long equal-power fades in use.

The Density knob allows significantly increasing how often grains are created, which combined with the Random knobs can result in a sound similar to Paulstretch, but without the stretch.  So just Paul, I suppose.

I added built-in documentation for the plugin so users can understand how to use it.

-----------------------

I have a quick demo of the plugin here (though it and its UI are a tad out of date):
https://www.youtube.com/watch?v=_8am4hjL-Tc

Rossmaxx commented on plugins/GranularPitchShifter/GranularPitchShifter.cpp:
```diff
@@ -0,0 +1,269 @@
+	const float d = dryLevel();
+	const float w = wetLevel();
+	
+	const ValueBuffer * pitchBuf = m_granularpitchshifterControls.m_pitchModel.valueBuffer();
```
Just so you know, valuebuffer is on it's way out in #7170. You don't need to worry for now. I'll intervene for fixing this PR up if that one gets merged.

Rossmaxx commented on plugins/GranularPitchShifter/GranularPitchShifter.cpp:
```diff
@@ -0,0 +1,269 @@
+	}
+	
+	const float shapeK = cosWindowApproxK(shape);
+	const int sizeSamples = static_cast<int>((m_sampleRate / size) * 0.5) * 2;
```
Do you really need to divide by 0.5, and multiply by 2? Wouldn't this cancel out?

LostRobotMusic replied:
No, it wouldn't cancel out because of the integer cast.  This forces sizeSamples to be an even number, which used to be required, but no longer is since waitMult was changed to a float.  I'll fix this.

Rossmaxx commented on plugins/GranularPitchShifter/GranularPitchShifter.cpp:
```diff
@@ -0,0 +1,269 @@
+		{
+			m_updatePitches = false;
+			
+			const double speed[2] = {std::exp2(m_truePitch[0] * (1. / 12.)), std::exp2(m_truePitch[1] * (1. / 12.))};
```
You use C array here and in next line, use std::array, is there a difference?

LostRobotMusic replied:
I use C arrays in all of my personal projects and sometimes forget to convert all of them over for PRs.

Rossmaxx commented on plugins/GranularPitchShifter/GranularPitchShifter.h:
```diff
@@ -0,0 +1,168 @@
+	// double index and fraction are required for good quality
+	inline float getHermiteSample(double index, int ch) {
+		const int index_floor = static_cast<int>(index);
+		const double fraction = index - index_floor;
```
Can you use const float here instead and avoid the casting at line 130?

LostRobotMusic replied:
No, this would very drastically reduce the plugin's audio quality (and wouldn't have any performance benefit since a cast would be required earlier anyway).  The input is already a double, and by waiting until after the integer part is removed to convert it to float, the resulting precision is significantly higher.  Without this, there are extremely ugly audible artifacts, especially when pitch shifting downward, since the float precision limitations would be applied while the index value is quite high.

Rossmaxx replied:
Alright.

Rossmaxx commented on plugins/GranularPitchShifter/GranularPitchShifterControlDialog.cpp:
```diff
@@ -0,0 +1,158 @@
+}
+
+
+QString GranularPitchShifterHelpView::s_helpText=
```
I believe it would help if this entire help document was it's own text file and we can read the contents at runtime if the user clicks the help button.

LostRobotMusic replied:
Reading those contents at runtime would require shipping that separate text file along with every LMMS copy (and then handling file paths and directory settings and all that), which really isn't necessary.  This help page is handled in the same way as X-Pressive currently.

Rossmaxx replied:
Can't you put that help file in along with the sources at this same folder. You won't have to deal with paths too.

LostRobotMusic replied:
I don't really see what the advantage in doing this would be.

Rossmaxx replied:
Just shifting lines to a seperate text file, if this doesn't please you, feel free to keep it as is.

Rossmaxx commented on plugins/GranularPitchShifter/GranularPitchShifterControlDialog.cpp:
```diff
@@ -0,0 +1,158 @@
+
+GranularPitchShifterHelpView::GranularPitchShifterHelpView():QTextEdit(s_helpText)
+{
+#if QT_VERSION < 0x50C00
```
```suggestion
#if (QT_VERSION < QT_VERSION_CHECK(5,12,0))
```

Rossmaxx reviewed (commented):
Just some minor stuff i came across.

Rossmaxx commented on plugins/GranularPitchShifter/GranularPitchShifter.cpp:
```diff
@@ -0,0 +1,269 @@
+		}
+		
+		// this stuff is computationally expensive, so we should only do it when necessary
+		if (m_updatePitches)
```
I believe pulling out this "computationally expensive" if block to a seperate function might please codefactor

LostRobotMusic replied:
We had a discussion about CodeFactor's complex method complaints on Discord with past PRs and have concluded that it's best to ignore it.  I don't think I'll ever be releasing a plugin that satisfies these.  It's a measurement of Cyclomatic complexity: https://en.wikipedia.org/wiki/Cyclomatic_complexity

Leaving this part in this function does not hinder readability in any way, and pulling it out of that function would result in 44100 unnecessary function calls per second, unless it's inlined in which case it would have to be placed in the header file, which _would_ hinder readability.

Veratil replied:
I really want to just disable codefactor, we ignore it anyway

tresf replied:
I've told codefactor to ignore this issue.  If this becomes a burden, we can ignore the entire file, or disable it project-wide.  Resolving.

Gabrielxd195 commented:
A small graphic detail is that there are inconsistencies in the letters, for example: "range" and "second" are in lowercase, when it should be "Range" and "Second"; While "LATENCY", TWITCH, DENSITY, SHAPE, PITCH, SPRAY, JITTER, SIZZE, GLIDE, FDBK and FDBK are capitalized. I don't know if they are doing this on purpose, but I think it would be better if each word starts in capital letters, like: "Stereo", "Prefilter". To make it more pleasing to the eye. Greetings 

(Google Translate)

Rossmaxx commented:
Now that you said it. I've noticed it just now. It would be better to keep the abbreviations (RND, FDBK and co) the same and apply changes to the rest, to bring in some contrast for abbreviated names.

tresf commented:
We don't yet have a style guideline for acceptance of plugins.  This is largely due to the fact that we've historically allowed theming of plugins and those themes make use of the background image for labels.

At a glance, this plugin uses:
* `Title` case for the sections
* `CAPS` for the knobs

SFXR uses this same technique.

The only inconsistencies I can find:
* `5 seconds`: We would write this as `5 Seconds` elsewhere
* `range`: Probably should be `Range` to be consistent with other sections.

sakertooth commented on plugins/GranularPitchShifter/GranularPitchShifterControlDialog.h:
```diff
@@ -0,0 +1,78 @@
+		static GranularPitchShifterHelpView instance;
+		return &instance;
+	}
+	static void finalize()
+	{
+	}
```
Can't this be removed?

sakertooth commented on plugins/GranularPitchShifter/GranularPitchShifterControls.cpp:
```diff
@@ -0,0 +1,157 @@
+			m_jitterModel.setRange(0.f, 2.f, 0.0001f);
+			break;
+		}
+		default: {}
```
I would put a `break` here instead.
```suggestion
		default:
		     break;
```

sakertooth commented on plugins/GranularPitchShifter/GranularPitchShifterControls.h:
```diff
@@ -0,0 +1,89 @@
+#include "EffectControls.h"
+#include "GranularPitchShifterControlDialog.h"
+
+constexpr float GPS_RANGE_SECONDS[5] = {5, 10, 40, 40, 120};
```
As previously mentioned, this isn't the naming convention we use for global constants. This can also be moved into an anonymous namespace in`GranularPitchShifterControls.cpp` I believe.

sakertooth commented on plugins/GranularPitchShifter/GranularPitchShifter.h:
```diff
@@ -0,0 +1,168 @@
+		
+		float v0, v1, v2, v3;
+
+		if (index_floor == 0) {
+			v0 = m_ringBuf[m_ringBuf.size() - 1][ch];
+		} else {
+			v0 = m_ringBuf[index_floor - 1][ch];
+		}
```
Could really just use a ternary here, but I don't mind if you want to leave it.

sakertooth reviewed (commented):
Tested it, cool plugin :+1:

LostRobotMusic commented on plugins/GranularPitchShifter/GranularPitchShifterControlDialog.cpp:
```diff
@@ -0,0 +1,158 @@
+}
+
+void GranularPitchShifterControlDialog::showHelpWindow() {
+	GranularPitchShifterHelpView::getInstance()->close();
+	GranularPitchShifterHelpView::getInstance()->show();
```
It forces the window to be brought to the top and focused if it's already been created.  Without calling `close`, if the window's already been created and is currently behind other windows, it'll be stuck there until you go dig it out.

LostRobotMusic commented on plugins/GranularPitchShifter/GranularPitchShifter.h:
```diff
@@ -0,0 +1,168 @@
+	std::array<double, 2> m_phaseSpeed;
+	std::array<double, 2> m_grainSpeed;
+	double m_phase;
+};
```
Resolved on Discord.  The link is saying:
`Consider putting every definition that is in an implementation source file into an unnamed namespace...`
and not:
`Consider putting every definition into an implementation source file within an unnamed namespace...`

sakertooth commented on plugins/GranularPitchShifter/GranularPitchShifterControls.h:
```diff
@@ -0,0 +1,87 @@
+ *
+ */
+
+#ifndef GRANULARPITCHSHIFTER_CONTROLS_H
+#define GRANULARPITCHSHIFTER_CONTROLS_H
```
```suggestion
#ifndef LMMS_GRANULAR_PITCH_SHIFTER_CONTROLS_H
#define LMMS_GRANULAR_PITCH_SHIFTER_CONTROLS_H
```

sakertooth commented on plugins/GranularPitchShifter/GranularPitchShifterControls.h:
```diff
@@ -0,0 +1,87 @@
+
+} // namespace lmms
+
+#endif
```
```suggestion
#endif // LMMS_GRANULAR_PITCH_SHIFTER_CONTROLS_H
```

sakertooth commented on plugins/GranularPitchShifter/GranularPitchShifterControlDialog.h:
```diff
@@ -0,0 +1,75 @@
+ *
+ */
+
+#ifndef GRANULARPITCHSHIFTER_CONTROL_DIALOG_H
+#define GRANULARPITCHSHIFTER_CONTROL_DIALOG_H
```
```suggestion
#ifndef LMMS_GRANULAR_PITCH_SHIFTER_CONTROL_DIALOG_H
#define LMMS_GRANULAR_PITCH_SHIFTER_CONTROL_DIALOG_H
```

sakertooth commented on plugins/GranularPitchShifter/GranularPitchShifterControlDialog.h:
```diff
@@ -0,0 +1,75 @@
+
+} // namespace lmms
+
+#endif
```
```suggestion
#endif // LMMS_GRANULAR_PITCH_SHIFTER_CONTROL_DIALOG_H
```

sakertooth reviewed (commented):
After these few changes I think its good to go :+1: (on my end at least)

Veratil commented on plugins/GranularPitchShifter/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+INCLUDE(BuildPlugin)
+
+BUILD_PLUGIN(granularpitchshifter GranularPitchShifterEffect.cpp GranularPitchShifterControls.cpp GranularPitchShifterControlDialog.cpp MOCFILES GranularPitchShifterControls.h GranularPitchShifterControlDialog.h EMBEDDED_RESOURCES *.png)
```
Switch to a little more standard CMake style
```suggestion
include(BuildPlugin)

build_plugin(granularpitchshifter
	GranularPitchShifterEffect.cpp
	GranularPitchShifterControls.cpp
	GranularPitchShifterControlDialog.cpp
	MOCFILES GranularPitchShifterControls.h GranularPitchShifterControlDialog.h
	EMBEDDED_RESOURCES *.png)
```

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterControls.cpp:
```diff
@@ -0,0 +1,158 @@
+	connect(&m_rangeModel, &ComboBoxModel::dataChanged, this, &GranularPitchShifterControls::updateRange);
+}
+
+void GranularPitchShifterControls::updateRange() {
```
I know we hate Allman, and I'm all for switching to K&R but for now we should stick with the Allman style for consistency.
```suggestion
void GranularPitchShifterControls::updateRange()
{
```

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterControls.cpp:
```diff
@@ -0,0 +1,158 @@
+	switch (m_rangeModel.value())
+	{
+		case 0:// 5 seconds
+		{
```
We could actually remove the braces since there's no variable scope, and deindent the case lines for less nested look.

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.cpp:
```diff
@@ -0,0 +1,269 @@
+	const float d = dryLevel();
+	const float w = wetLevel();
+	
+	const ValueBuffer * pitchBuf = m_granularpitchshifterControls.m_pitchModel.valueBuffer();
+	const ValueBuffer * pitchSpreadBuf = m_granularpitchshifterControls.m_pitchSpreadModel.valueBuffer();
```
```suggestion
	const ValueBuffer* pitchBuf = m_granularpitchshifterControls.m_pitchModel.valueBuffer();
	const ValueBuffer* pitchSpreadBuf = m_granularpitchshifterControls.m_pitchSpreadModel.valueBuffer();
```

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.cpp:
```diff
@@ -0,0 +1,269 @@
+		{
+			m_updatePitches = false;
+			
+			std::array<double, 2> speed = {std::exp2(m_truePitch[0] * (1. / 12.)), std::exp2(m_truePitch[1] * (1. / 12.))};
+			std::array<double, 2> ratio = {speed[0] / m_speed[0], speed[1] / m_speed[1]};
```
A little easier to read
```suggestion
			std::array<double, 2> speed = {
				std::exp2(m_truePitch[0] * (1. / 12.)),
				std::exp2(m_truePitch[1] * (1. / 12.))
			};
			std::array<double, 2> ratio = {
				speed[0] / m_speed[0],
				speed[1] / m_speed[1]
			};
```

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.cpp:
```diff
@@ -0,0 +1,269 @@
+			if (spray > 0)
+			{
+				sprayResult[0] = (fast_rand() / static_cast<float>(FAST_RAND_MAX)) * spray * m_sampleRate;
+				sprayResult[1] = linearInterpolate(sprayResult[0],
+									(fast_rand() / static_cast<float>(FAST_RAND_MAX)) * spray * m_sampleRate,
+									spraySpread);
```
Less indent heavy here
```suggestion
				sprayResult[1] = linearInterpolate(
					sprayResult[0],
					(fast_rand() / static_cast<float>(FAST_RAND_MAX)) * spray * m_sampleRate,
					spraySpread
				);
```

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.cpp:
```diff
@@ -0,0 +1,269 @@
+			for (int i = 0; i < 2; ++i)
+			{
+				readPoint[i] = m_writePoint - latency - sprayResult[i];
+				if (readPoint[i] < 0) {readPoint[i] += m_ringBufLength;}
```
```suggestion
				if (readPoint[i] < 0) { readPoint[i] += m_ringBufLength; }
```

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.cpp:
```diff
@@ -0,0 +1,269 @@
+			}
+			
+			std::array<int, 2> readPoint;
+			int latency = std::max(int(std::max(sizeSamples * (std::max(m_speed[0], m_speed[1]) * grainSpeed - 1.), 0.) + SafetyLatency), minLatency);
```
What's this `int()` doing in there? `static_cast<int>()`?

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.h:
```diff
@@ -0,0 +1,168 @@
+		
+		float v0, v1, v2, v3;
+
+		if (index_floor == 0) {
+			v0 = m_ringBuf[m_ringBuf.size() - 1][ch];
+		} else {
+			v0 = m_ringBuf[index_floor - 1][ch];
+		}
```
Another K&R snuck in I'm sorry 😭but we can at least do this
```suggestion
		if (index_floor == 0) { v0 = m_ringBuf[m_ringBuf.size() - 1][ch]; }
		else { v0 = m_ringBuf[index_floor - 1][ch]; }
```

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.h:
```diff
@@ -0,0 +1,168 @@
+		
+		v1 = m_ringBuf[index_floor][ch];
+		
+		if(index_floor >= m_ringBuf.size() - 2) {
+			v2 = m_ringBuf[(index_floor + 1) % m_ringBuf.size()][ch];
+			v3 = m_ringBuf[(index_floor + 2) % m_ringBuf.size()][ch];
+		} else {
```
```suggestion
		if (index_floor >= m_ringBuf.size() - 2)
		{
			v2 = m_ringBuf[(index_floor + 1) % m_ringBuf.size()][ch];
			v3 = m_ringBuf[(index_floor + 2) % m_ringBuf.size()][ch];
		}
		else
		{
```

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.h:
```diff
@@ -0,0 +1,168 @@
+	
+	// adapted from signalsmith's crossfade approximation:
+	// https://signalsmith-audio.co.uk/writing/2021/cheap-energy-crossfade
+	float cosHalfWindowApprox(float x, float k) {
```
```suggestion
	float cosHalfWindowApprox(float x, float k)
	{
```

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.h:
```diff
@@ -0,0 +1,168 @@
+		return C * C;
+	}
+	// 1-2 fades between equal-gain and equal-power
+	float cosWindowApproxK(float p) {
+		return -6.0026608f + p * (6.8773512f - 1.5838104f * p);
+	}
```
```suggestion
	float cosWindowApproxK(float p) { return -6.0026608f + p * (6.8773512f - 1.5838104f * p); }
```

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.h:
```diff
@@ -0,0 +1,168 @@
+		void setCoefs(float sampleRate, float cutoff)
+		{
+		    const float g = std::tan(F_PI * cutoff / sampleRate);
+		    const float ginv = g / (1.0f + g * (g + 1.414213562));
```
Maybe we should magic const the sqrt2 here instead of hardcoding the number. If we use it elsewhere in code it might be useful to throw it in `lmms_math.h` or `lmms_constants.h`.

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.h:
```diff
@@ -0,0 +1,168 @@
+			readPoint{readPointL, readPointR},
+			phaseSpeed{phaseSpeedL, phaseSpeedR},
+			grainSpeed{grainSpeedL, grainSpeedR},
+			phase{0} {}
```
A little hard to see, just make it more obvious.
```suggestion
			phase{0}
		{}
```

Veratil commented on plugins/GranularPitchShifter/GranularPitchShifterEffect.h:
```diff
@@ -0,0 +1,168 @@
+	}
+	
+	// double index and fraction are required for good quality
+	float getHermiteSample(double index, int ch) {
```
```suggestion
	float getHermiteSample(double index, int ch)
	{
```

DomClark commented on plugins/GranularPitchShifter/artwork.png:
I hate to be the one who hates fun, but can we please try to keep the contents of this repository vaguely sensible and professional?

LostRobotMusic replied:
When squash-and-merge is used, only the resulting squashed commit appears in the repository, not the individual commits that were part of the pull request.

Rossmaxx replied:
Dom knows it better than all of us. He's just requesting some level of professionalism.

LostRobotMusic replied:
Oh wait, I didn't notice the file this review was on, I thought you were talking about the commit messages.

The answer is no.  Unless you really really want to, but then I will be sad and probably cry.

Rossmaxx replied:
Ohh i too thought about the commit messages.

LostRobotMusic commented:
![image](https://github.com/LMMS/lmms/assets/34612565/ec306bc7-f384-4e6d-97e7-2c767e470221)

LostRobotMusic commented:
I added a 7 Hz 1-pole highpass to the feedback loop to prevent DC offset buildups.  I also added a very CPU-efficient safety saturator I made to guarantee the feedback is always safe.  With a sufficiently low input volume (below +24 dBFS), all it needs is an `abs` and a correctly predicted branch.

This plugin is 100% ready for merge as far as I can tell, but it needs one more approval first.
